### PR TITLE
Complex balancing

### PR DESCRIPTION
**Pull request by vyudu**
## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API
  
## Additional context

A first pass at the complex-balance implementation. Does not seem like Graphs.jl has a matrix-tree implementation, so I have the matrix tree implementation here for now, but I'll work on adding it as a PR to Graphs.jl (and in particular implementing a method of generating the spanning trees that doesn't simply generate every combination of n-1 edges).

### CODE REVIEW

**Review by isaacsas — CHANGES_REQUESTED** [review]
Let me start by saying this is a great first PR!

Here are a first set of comments -- these are more general Julia practices / Catalyst internals comments than method specifc or performance specific. 

I need to finish going through the paper and then I will give a second review more focused on the algorithmic details, but you can at least address these for now.

**Comment by TorkelE** [comment]
I second that GitKraken is a really nice tool, if you haven't checked it out yet.

File: src/networkapi.jl
Comment (vyudu) [reply src/networkapi.jl]: Ah, this was an artifact of a previous implementation that I forgot to delete. Good to know about those two functions, though, will implement the latter check

File: src/networkapi.jl
Comment (vyudu) [reply src/networkapi.jl]: I think I made this change correctly, but let me know. Fortunately it seemed that I don't actually need to use the fixed parameter vector directly, just the mapping.

File: test/network_analysis/network_properties.jl
Comment (vyudu) [reply test/network_analysis/network_properties.jl]: Ditto on this - let me know if how I'm constructing the parameter maps in the test file in the new commit is correct.

File: Project.toml
Comment (isaacsas) [inline Project.toml:18]:
```suggestion
```
Can you remove the MetaGraphs dependency here?

File: src/networkapi.jl
Comment (isaacsas) [reply src/networkapi.jl:1504]: Also, currently `ismassaction` would still declare that something like `k1*k2, A --> B` is mass action, but this would break your code here. After checking all reactions are mass action, what you need to do is something like
```julia
ratevalues = [substitute(rate, rates) for rate in reactionrates(rs)]
```
This should numerically evaluate `k1*k2` using the substitution vector/dict the user provided. It isn't the most efficient way to do this, but is fine for the first version of this code.

File: src/networkapi.jl
Comment (isaacsas) [inline src/networkapi.jl:1496]:
```suggestion
```
See below -- this isn't how we should do the check I think since even for mass action reactions rates may not be just a single symbolic.

File: src/networkapi.jl
Comment (isaacsas) [inline src/networkapi.jl:1493]: This should work for Vector/tuple mappings too as we commonly suggest them for users, along with specifying `Symbols` for the first component. You can use `symmap_to_varmap` to convert `Symbol` mappings to symbolic mappings, see https://docs.sciml.ai/Catalyst/stable/api/catalyst_api/#Catalyst.symmap_to_varmap.

File: src/networkapi.jl
Comment (isaacsas) [inline src/networkapi.jl:1563]:
```suggestion
```

File: src/networkapi.jl
Comment (isaacsas) [inline src/networkapi.jl:1568]:
```suggestion
```
See comments above about how to numerically calculate the rates (I'd suggest validating and calculating them above, and then just passing the final numeric rate vector into this function. It may help speedup this function by passing in a more concrete type for `rates`, i.e. one that doesn't depend on `Any`).

File: src/networkapi.jl
Comment (isaacsas) [inline src/networkapi.jl:1578]:
```suggestion
        s = findfirst(==(-1), @view D[:,r])
        p = findfirst(==(1), @view D[:,r])
```
This avoids allocating a copy vector for each column.

File: src/networkapi.jl
Comment (isaacsas) [reply src/networkapi.jl:1493]: I would just not type `rates` for simplicity, but in the docstring mention the input options (vectors, tuples, dict mappings as accepted by ModelingToolkit for specifying parameter values).

File: src/networkapi.jl
Comment (isaacsas) [inline src/networkapi.jl:1485]:
```suggestion
    iscomplexbalanced(rs::ReactionSystem, rates)
```

File: src/networkapi.jl
Comment (isaacsas) [inline src/networkapi.jl:1637]: This should go before your added code since it goes with the unit validation above.

**Review by isaacsas — CHANGES_REQUESTED** [review]
Can you add one more test that uses a network that is complex balanced with multiple nonlinearities and non-trivial rate expressions (i.e. reactions like `k1*k2, 3A + 2B --> 3*C`)? Just to stress the code a bit more.

In terms of performance optimization I think that can be a followup. So I'm happy to merge once you address this last set of comments.

File: src/networkapi.jl
Comment (isaacsas) [reply src/networkapi.jl:1485]: Also, probably it is more descriptive to call the second argument `parametermap` or similar as it should be a mapping giving numeric values for all system parameters.

File: src/networkapi.jl
Comment (isaacsas) [reply src/networkapi.jl:1493]: Inside the function you are welcome to convert the inputted mapping to a dictionary for ease of use if you want (and needed). It might make sense to just have three dispatches
```julia
function iscomplexbalanced(rs::ReactionSystem, parametermap::Vector)
     pdict = Dict(parametermap)
     iscomplexbalanced(rs, pdict)
end

function iscomplexbalanced(rs::ReactionSystem, parametermap::Tuple)
     pdict = Dict(parametermap)
     iscomplexbalanced(rs, pdict)
end

function iscomplexbalanced(rs::ReactionSystem, parametermap::Dict)
    your existing code
end

iscomplexbalanced(rs::ReactionSystem, parametermap) = error("parametermap is not in an allowable format.)
```
or similar.

File: src/networkapi.jl
Comment (isaacsas) [inline src/networkapi.jl:1612]:
```suggestion
    # constructed rooted trees for every vertex, compute sum
```

File: src/networkapi.jl
Comment (vyudu) [reply src/networkapi.jl:1634]: Folded this into a future PR

File: src/networkapi.jl
Comment (vyudu) [reply src/networkapi.jl:1496]: I've changed it to the following - does this solve this concern? 

```julia
if length(parametermap) != numparams(rs) error... end
```

Comment (isaacsas) [reply src/networkapi.jl:1496]: Sure, that is fine to check.

**Comment by isaacsas** [comment]
CI is still failing?

**Comment by isaacsas** [comment]
This looks like it might have been rebased on an older version of master?

**Comment by isaacsas** [comment]
(Generally I find it easier to merge instead of rebase to update a PR.)

**Comment by TorkelE** [comment]
@vyudu I tried to merge this onto master for you, but couldn't get it to work unfortunately. If you check the "Files changed" tab at the top of the PR you see that there are a very large number of changes:
image

What probably have happened is that this was a PR of some branch, and then it got rebased on an updated master, and somewhere some of the changes of the branches it originally was branched off of was carried through when this one was rebased on the master again.

Since you are developing on your own master branch, I would try (as Sam pointed out here, and to me earlier) to simply merge your master branch with Catalyst's master branch, and then push the PR again. If that does not work out, and the changes introduced here are very straightforward (i.e. just adds functions at the bottom of the network_analysis.jl file, and some tests) it might be easiest to simply copy these into a new, clean, branch and make a new PR from that.

File: src/Catalyst.jl
Comment (isaacsas) [inline src/Catalyst.jl:12]:
```suggestion
using LinearAlgebra, Combinatorics
using JumpProcesses: JumpProcesses, JumpProblem, 
                     MassActionJump, ConstantRateJump, VariableRateJump,
                     SpatialMassActionJump
```

**Comment by isaacsas** [comment]
Thanks @vyudu great first PR.

**Comment by TorkelE** [comment]
> Thanks @vyudu great first PR.

Second that! And sorry for giving you poor advice on the rebase/merge thing. I tried @isaacsas's method properly now as well and it works great

**Pull request merged**
